3 단계 - 자동차 경주

Jinni01 opened the pull request.

Jinni01 committed:
feat: Step2 feedback 반영 및 개선

- enum 리플렉션반환 코드 제거
- 코틀린 collection api 활용(zip, fold)

Jinni01 committed:
feat: Step2 feedback 반영 및 개선

- 불필요 주석 제거

Jinni01 committed:
feat: 레이싱카 객체 이동 테스트 케이스 작성

Jinni01 committed:
feat: RacingCarPosition 도메인 분리

Jinni01 committed:
feat: racingCar attemptMove 메서드 추가

Jinni01 committed:
feat: RaceStrategy 인터페이스 및 RandomStrategy 싱글턴 객체 추가

Jinni01 committed:
feat: RaceContainer 및 RacingCarFactory 클래스 추가

- ErrorMessage 클래스 추가

Jinni01 committed:
feat: RacingCarFactory 싱글턴으로 변경

Jinni01 committed:
feat: Race클래스 추가, 스텝/전체 기록 남기는 로직 추가

Jinni01 committed:
feat: InputView/ResultView 클래스 추가, main 함수 작성

lhs0610 commented on src/main/kotlin/application/KotlinCarRacingApplicaiton.kt:
```diff
@@ -0,0 +1,18 @@
+    val racingcarQuantity = InputView.inputRacingCarQuantity()
+    val totalStep = InputView.inputTotalStep()
+
+    val race = Race(racingcarQuantity = racingcarQuantity, totalStep = totalStep, RandomStrategy)
```
RandomStrategy만 named mapping을 안해주는 이유가 있을까요?

Commit:
feat : RaceHistoryDto data class 변경

Commit:
feat : code lint

lhs0610 commented on src/main/kotlin/domain/racingcar/RacingCar.kt:
```diff
@@ -0,0 +1,21 @@
+class RacingCar(
+    private val strategy: RaceStrategy = RandomStrategy
+) {
+    private val _position = RacingCarPosition()
```
위치정보를 내부 프로퍼티로 지정해주셨네요. 
이런 경우 처음부터 위치를 0이 아닌 3이나 5로 지정하고 싶을 경우 제약이 생길 것 같아요. 
해당 객체가 불변이 아닌 가변 값 객체로 사용되기에 위변조를 막기 위해 이과 같이 한 것 같은데, 값 객체를 불변으로 만들면 어떨까요 그럼 val이 아닌 var로 지정해 생성자 파라미터로 받아도 문제가 안 될 것 같군요

lhs0610 commented on src/main/kotlin/domain/racingcar/RacingCar.kt:
```diff
@@ -0,0 +1,21 @@
+    val position: Int
+        get() = _position.position
+
+    fun attemptMove() {
+        move(strategy.move())
+    }
+
+    fun move(distance: Int) {
+        _position.move(distance)
+    }
```
둘 다 public이네요. 외부에서 그냥 move(Int.MAX_VALUE)로 이동할수도 있을 것 같네요.

lhs0610 commented on src/main/kotlin/domain/racingcar/RacingCarContainer.kt:
```diff
@@ -0,0 +1,25 @@
+    private val racingCars: List<RacingCar>
+
+    init {
+        require(racingCarQuantity > 0) { ErrorMessage.INVALID_RACINGCAR_QUANITY.message }
```
매직 넘버는 상수로 분리해주면 가독성을 높힐 수 있을 것 같네요.

lhs0610 commented on src/main/kotlin/domain/racingcar/RacingCarContainer.kt:
```diff
@@ -0,0 +1,25 @@
+        racingCars.forEach { it.attemptMove() }
+        val positions = racingCars.map { it.position }.toList()
+
+        return StepHistoryDto(positions)
```
해당 컨테이너는 도메인인데 DTO를 의존하고 있네요 🤔

lhs0610 commented on src/main/kotlin/domain/racingcar/RacingCarContainer.kt:
```diff
@@ -0,0 +1,25 @@
+
+    fun stepProgress(): StepHistoryDto {
+        racingCars.forEach { it.attemptMove() }
+        val positions = racingCars.map { it.position }.toList()
```
직접 map으로 꺼내기보다 racingCars.getAllPositions()처럼 요청으로 응답을 받아도 될 것 같아요

lhs0610 commented on src/main/kotlin/domain/racingcar/RacingCarPosition.kt:
```diff
@@ -0,0 +1,10 @@
+package domain.racingcar
+class RacingCarPosition(
+    private var _position: Int = 0
```
생성자 파라미터에서 밑줄을 사용하는건 권장되지 않아요. 현재 코드를 보면 `_position`과 `position` 두 필드가 인스턴스내에 동시에 존재하고 있어요. java로 치면

```java
class RacingCarPosition {
    private Int _position = 0;
    private final int position

    public RacingCarPosition(int _position) {
        this._position = _position;
        this.position = this._position;
    }

}

```

위와 같은 코드인 것입니다.

lhs0610 commented on src/main/kotlin/domain/racingcar/component/Race.kt:
```diff
@@ -0,0 +1,27 @@
+import domain.racingcar.strategy.RaceStrategy
+import domain.racingcar.strategy.RandomStrategy
+
+class Race(
```
해당 객체의 계층위치가 적절한지 고민할 필요가 있습니다. 
컴포넌트? 서비스 계층으로 보이는데 패키지 위치는 도메인에 있네요. 

도메인계층과 서비스계층의 차이가 무엇일까요? 
책임지는 범위와 반환 되어 전달해주는 계층의 차이가 있죠. 
또한 도메인과 다르게 서비스 계층은 보통 상태를 갖지 않고 재사용성을 제공한다는 특징도 있어요. 

해당 객체는 그럼 무슨 계층이고 무엇을 의도한 객체인가요? 
제가 볼 때는 서비스 계층으로 보이는데, 그럼 서비스 계층의 특징인 무상태성을 가지고 있는것인지 의문입니다.

코틀린을 사용한다고 자바와 아키텍처적으로 다른 패러다임을 지향하는게 아닙니다. 
몇 가지 문법적 슈가 코드가 추가되어서 안티패턴들이 권장사양으로 바뀐정도지 전체적인 맥락은 달라지지 않았어요.

lhs0610 commented on src/main/kotlin/domain/racingcar/component/Race.kt:
```diff
@@ -0,0 +1,27 @@
+
+        val racingCarContainer = RacingCarContainer(racingcarQuantity, raceStrategy)
+
+        for (i in 1..totalStep) {
```
```suggestion
        (0 until totalStep).map { racingCarContainer.stepProgress() }
```
한 줄로 작성할수도 있을 것 같네요. 더 합치자면

```kotlin
        (0 until totalStep).map { racingCarContainer.stepProgress() }.let(::RaceHistoryDto)
```

이렇게도 표현 가능하겠네요

lhs0610 commented on src/main/kotlin/domain/racingcar/dto/RaceHistoryDto.kt:
```diff
@@ -0,0 +1,4 @@
+package domain.racingcar.dto
+
+class RaceHistoryDto(val stepHistories: List<StepHistoryDto>) {
```
dto면 data class를 쓰면 어떨까요? 그리고 중괄호 블럭을 제거해도 무관해보이네요

lhs0610 commented on src/main/kotlin/domain/racingcar/strategy/RandomStrategy.kt:
```diff
@@ -0,0 +1,17 @@
+
+    private const val MIN_MOVE = 4
+    private const val STAY = 0
+    override fun move(): Int {
```
경주의 요구조건은 4이상일 경우 움직일 수 있다지 4이상일 경우 random point만큼 이동할수 있다가 아닙니다.

lhs0610 commented on src/main/kotlin/utilities/NumberUtil.kt:
```diff
@@ -0,0 +1,16 @@
+
+import enums.ErrorCode
+
+class NumberUtil {
```
xxxUtil클래스를 만드는건 곧 xxx에 관련된 모든 책임을 다 가져가겠다는 것이기에, 일반적으로 정말 필요해지는 상황이 아니라면 생성에 제약을 두는게 좋습니다. 
한두군데서 중복된다고 Util클래스를 만들었다간 나중에 이 클래스에 대한 유지보수성을 책임지기 힘들어져요. 

클린코드의 10장 클래스쪽을 참고해보셔도 좋을 것 같아요.

lhs0610 commented on src/test/kotlin/racingcar/RacingCarTest.kt:
```diff
@@ -0,0 +1,20 @@
+import io.kotest.core.spec.style.ExpectSpec
+import io.kotest.matchers.shouldBe
+
+class RacingCarTest: ExpectSpec({
```
테스트가.......하나뿐이네요?

lhs0610 reviewed (changes requested):
의진님 Step3 레이싱 미션을 드디어 시작하셨네요. 

전체적으로, 코틀린의 코드를 경험해보려고 여러 시도를 해보셨어요 👍 
그런데, 코틀린은 몇몇 슈가 코드를 제공해서 코드의 작성의 이점을 줄 수는 있지만 아키텍처 레벨에서 자바와 차이를 두지는 않아요. 그래서 이펙티브 자바에서도 말했던 내용들이 대부분 동일하게 적용되고 클린코드나 클린 아키텍처의 내용도 그대로 유지됩니다. 

그런 부분들을 고려해서 피드백을 남겨드렸어요. 
추가적으로 현재 테스트코드가 아예작성되지 않은 기능단위 개발이 된 것 같아요. 
이 부분도 좀 확인해주시면 좋을 것 같네요. 

피드백 확인 후 다시 리뷰요청 부탁드려요

Commit:
refactor: 프로젝트 패키지 구조 스텝별 분리 및 개선

- 각각의 스텝을 루트로 형상 가져가도록 수정

Commit:
feat: 자동차 경주 요구 사항 정의서 작성

Commit:
feat: RacingCarPosition value class화, 테스트 케이스 추가

- 빌드 에러 발생시키는 미완성 클래스들 제거

Commit:
feat: RacingCarPosition 뺴기 연산 테스트 케이스 추가

Commit:
feat: RacingCar 클래스 리팩토링, 테스트 케이스 추가

- position 주입 가능하고, 참조 변경할 수 있도록 수정
- 바뀐 전략에 맞춘 attemptMove 메서드 작성

Commit:
feat: RandomStrategy 클래스 리팩토링

- 바뀐 인터페이스 정책에 따라 내부 구현 수정

Commit:
feat: RacingCar 클래스 수정

- 내부에서 정책을 소유하는 것이 아닌, 이동시 메서드에서 인자로 받도록 수정
- 연관된 테스트 케이스 코드 수정

Commit:
feat: RacingCarFactory 클래스 리팩토링

- 레이싱카 리스트 생성 메서드에서 정책을 주입받지 않도록 수정(레이싱카 이동 메서드 변경 관련 반영)
- 생성될 레이싱카들의 초기 위치 함수 파라미터 추가

Commit:
feat: RacingCarContainer 클래스 리팩토링

- 생성될 레이싱카들의 초기 위치 함수 파라미터 추가
- StepHistoryDto 도메인 객체에 의존하지 않고 정수 결과값 리스트 반환하도록 수정
- List<RacingCar> 타입의 확장 함수 추가

Commit:
feat: RacingCarContainer 클래스 테스트 케이스 작성

- inspector 활용

Commit:
feat: Race 클래스 리팩토링, 테스트 케이스 작성

- start 메서드 로직 수정, 코틀린 문법 활용

Commit:
feat: 체크리스트 반영

Commit:
feat: GameController / GameService 싱글턴 클래스 추가

Commit:
feat: 입출력 관련 작업 추가, 각 동작 연결

lhs0610 commented on src/test/kotlin/racingcar/RacingCarTest.kt:
```diff
@@ -1,20 +1,81 @@
         }
     }
+
+    context("RacingCarFactory test") {
+        expect("generate racingcar list test") {
+            val racingCars = RacingCarFactory.generateRacingCars(3, RacingCarPosition(0))
+            for (racingCar in racingCars) {
+                racingCar.getDistance() shouldBe 0
+            }
+        }
+    }
+
+    context("RacingCarContainer test") {
+        expect("step progress test") {
+            val racingCarContainer = RacingCarContainer(5, RacingCarPosition(0))
+            val stepHistory = racingCarContainer.stepProgress { true }
+
+            stepHistory.shouldHaveSize(5)
+            stepHistory.forAll { it.shouldBe(1) }
+        }
+    }
+
+    context("Race test") {
+        expect("start test") {
+            val race = Race(7, 2) { true }
+            val raceHistory = race.start()
+
+            raceHistory.stepHistories.shouldHaveSize(2)
+            raceHistory.stepHistories.last().positions.forAll { it.shouldBe(2) }
+        }
+    }
```
테스트 케이스가 대상으로 하고 있는 대상이 각기 다르네요. 테스트 클래스 네이밍 관례도 얘기를 했던 것 같은데, 
인수테스트가 아닌 이상 하나의 테스트 클래스에서는 하나의 대상에 대한 테스트만 해주는게 좋을 것 같네요.

Jinni01 replied:
클래스별로 분리해서 만들 수 있도록 하겠습니다. expect스펙의  context를 활용해서 구분해보려 했는데 나누는 편이 나을 것 같네요

lhs0610 commented on src/main/kotlin/racingcar/component/Race.kt:
```diff
@@ -0,0 +1,27 @@
+    private val raceStrategy: RaceStrategy
+) {
+
+    fun start(): RaceHistoryDto {
+        val racingCarContainer = RacingCarContainer(racingcarQuantity, RacingCarPosition(START_DISTANCE))
+
+        return (0 until totalStep).map {
+            val stepResult = racingCarContainer.stepProgress(raceStrategy)
+            StepHistoryDto(stepResult)
+        }.let(::RaceHistoryDto)
+    }
```
Race 객체를 아예 도메인으로 결정하신 것 같아요. 
그런데 start로직을 보면 자기 자신에 대한 상태 핸들링이 아니라 자동차 목록을 직접 만들고, 경주를 라운드별로 진행하고 있는데, 이게 RaceService와 Race객체의 차이가 어디에 있을까요?

lhs0610 commented on src/main/kotlin/racingcar/component/RacingCarFactory.kt:
```diff
@@ -0,0 +1,13 @@
+object RacingCarFactory {
+    fun generateRacingCars(quantity: Int, position: RacingCarPosition): List<RacingCar> {
+
+        val racingCars = mutableListOf<RacingCar>()
+        for (i in 1..quantity) { racingCars.add(RacingCar(position)) }
+        return racingCars.toList()
```
```suggestion
       return List(quantity) {
           RacingCar(position)
       }
```

이렇게도 표현 가능할 것 같네요

lhs0610 commented on src/main/kotlin/racingcar/controller/GameController.kt:
```diff
@@ -0,0 +1,16 @@
+
+    fun startRacingGame() {
+        val userInput = InputView.getUserInput()
+        val result = GameService.startRacingGame(userInput.racingcarQuantity, userInput.totalStep, RandomStrategy)
```
파라미터가 2개 이상일 경우 리팩토링 대상으로 지정하거나 named parameter를 활용해보면 좋을 것 같네요.

Jinni01 replied:
startRacingRame 메서드의 1, 2 번째 파라미터 타입이 Int로 동일해 오용의 여지가 있네요. named argument(parameter)를 잘 활용해서 구분하도록 하겠습니다.

lhs0610 commented on src/main/kotlin/racingcar/domain/RacingCar.kt:
```diff
@@ -0,0 +1,23 @@
+    }
+
+    fun attemptMove(raceStrategy: RaceStrategy) {
+        if (!raceStrategy.allowMove()) {
+            return
+        }
+        move()
```
```suggestion
        if (raceStrategy.allowMove()) {
            move()
        }
```

Jinni01 replied:
긴 코드에서 상단부 반환을 쓰는 습관을 간단한 코드에서도 불필요하게 적용한 케이스인 것 같네요. 유연하게 적용할 수 있도록 하겠습니다.

lhs0610 commented on src/main/kotlin/racingcar/domain/RacingCar.kt:
```diff
@@ -0,0 +1,23 @@
+    }
+
+    private fun move() {
+        position += 1
```
매직넘버는 상수로

lhs0610 commented on src/main/kotlin/racingcar/domain/RacingCarContainer.kt:
```diff
@@ -0,0 +1,25 @@
+import racingcar.constant.Message
+import racingcar.strategy.RaceStrategy
+
+class RacingCarContainer(
+    racingCarQuantity: Int,
+    position: RacingCarPosition
+) {
+    private val racingCars: List<RacingCar>
+
+    init {
+        racingCars = RacingCarFactory.generateRacingCars(racingCarQuantity, position)
+    }
```
- 팩토리 함수를 고려해봐도 좋을 것 같아요. 클래스 시그니처와 생성자 프로퍼티만 봐서는 이 객체가 무엇을 의도하고 보여주는지 알기 힘듭니다.
- 생성자 파라미터로 자동차대수를 입력하고 위치값을 다 입력받고 내부적으로 알아서 생성하게 해버리면, 실질적으로 내부의 자동차의 각 포지션정보들도 모두 제어할 수 없는 영역이 되겠죠? 실제로 레이싱을 할 때는 자동차별로 시작 위치가 다른데 그런 부분들에 대한 고려까지 한다면 해당 로직은 사용하기 힘들 것 같네요.

lhs0610 commented on src/main/kotlin/racingcar/view/ResultView.kt:
```diff
@@ -0,0 +1,27 @@
+
+        stepHistory.positions.forEach() {
+            for (i in 1..it) {
+                print(FOOTPRINTS_SYMBOL)
```
it의 크기가 10,000개 이상일 경우 print를 매 번 새로 호출하는것만으로 퍼포먼스 문제가 있을 것 같지 않나요?

Jinni01 replied:
출력 인터페이스쪽에서 성능 부분을 덜 고려했네요. 완성된 문자열을 줄마다 출력할 수 있도록 하겠습니다

lhs0610 commented on src/test/kotlin/racingcar/RacingCarTest.kt:
```diff
@@ -1,20 +1,81 @@
+        expect("construct test") {
+            val racingCar = RacingCar(RacingCarPosition(0))
+
+            racingCar.getDistance() shouldBe 0
```
테스트가 실패하네요. 연산자 오버로딩이 잘 못 구현된 것 같네요

Jinni01 replied:
minus 테스트쪽에 달아주신 코멘트였군요. 연산자 만들 때 기존 plus연산 내용을 일부 복사해서 썼는데 코드 작업 과정에서 잘못 수정한 것 같습니다. 제대로 테스트 돌아갈 수 있도록 수정하겠습니다.

lhs0610 reviewed (changes requested):
의진님 피드백 반영 잘 해주셨어요. 
추가적인 피드백을 남겨드렸는데, 현재 테스트가 실패하는게 있고, 
klint 도 지금 실패하고 있어요.. 항상 klint 체크 후 PR 제출해주세요.

수정 이후 다시 리뷰요청 부탁드려요

Commit:
refactor: request change 내용 반영 및 개선

- 테스트 대상 클래스별로 테스트클래스 작성(분리)
- 실패 테스트 케이스에 대한 로직 수정
- 출력문 로직 개선(결과 문자열 빌드)
- Factory/Container 객체에서 레이싱카 설정 정보를 통해 차별화할 수 있도록함
- 불필요 전위 반환 로직 수정
- named argument 활용
- Race 클래스 제거 및 RaceGameService 로직 통합

Commit:
refactor: request change 내용 반영 및 개선

- requirements.md 변경 내용 반영

lhs0610 commented on src/main/kotlin/racingcar/view/ResultView.kt:
```diff
@@ -17,11 +17,12 @@ object ResultView {
-            }
-            println()
-        }
+        val footPrints = stepHistory.positions.map {
+            val sb = StringBuilder()
+            repeat(it) { sb.append(FOOTPRINTS_SYMBOL) }
+            sb.toString()
+        }.toList()
```
stringBuilder DSL을 사용해도 좋습니다 👀

lhs0610 commented on src/main/kotlin/racingcar/service/RaceGameService.kt:
```diff
@@ -0,0 +1,21 @@
+
+    private const val START_DISTANCE = 0
+
+    fun startRacingGame(racingcarQuantity: Int, totalStep: Int, raceStrategy: RaceStrategy): RaceHistoryDto {
```
해당 로직에 대한 테스트 케이스가 존재하지 않는 것 같아요

Jinni01 replied:
앗 서비스로직 테스트를 빼먹었네요. 테스트케이스 추가할 수 있도록 하겠습니다.

lhs0610 reviewed (approved):
의진님 피드백 반영 👍 
서비스 클래스 테스트 케이스가 누락된 것 같은데 그 부분만 챙겨주시면 좋을 것 같고, 
이만 merge하도록 하겠습니다 4단계 진행 해주세요.
